File: android/app/src/main/java/com/mlbbhub/app/VoidContactsPlugin.java
package com.mlbbhub.app;

import android.Manifest;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.provider.ContactsContract;

import androidx.core.content.ContextCompat;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

import java.util.HashSet;
import java.util.Set;

/**
 * Read-only contact lookup by name — lets VOID resolve "Mom" or "Kassie" to
 * a phone number so it can hand off to WhatsApp/SMS with the chat already
 * open and the message pre-filled. Never writes to or modifies contacts.
 */
@CapacitorPlugin(
    name = "VoidContacts",
    permissions = {
        @Permission(strings = { Manifest.permission.READ_CONTACTS }, alias = "contacts")
    }
)
public class VoidContactsPlugin extends Plugin {

    private boolean hasPermission() {
        return ContextCompat.checkSelfPermission(getContext(), Manifest.permission.READ_CONTACTS)
            == PackageManager.PERMISSION_GRANTED;
    }

    @PluginMethod
    public void isEnabled(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("value", hasPermission());
        call.resolve(ret);
    }

    @PluginMethod
    public void requestAccess(PluginCall call) {
        if (hasPermission()) { call.resolve(); return; }
        requestAllPermissions(call, "contactsPermCallback");
    }

    @PermissionCallback
    private void contactsPermCallback(PluginCall call) {
        if (hasPermission()) call.resolve();
        else call.reject("CONTACTS_PERMISSION_DENIED");
    }

    /** query: a name (or partial name) to search for. Returns up to 5
     *  name+phone matches, best (alphabetically first) match first. */
    @PluginMethod
    public void find(PluginCall call) {
        if (!hasPermission()) {
            JSObject ret = new JSObject();
            ret.put("ok", false);
            ret.put("error", "NOT_ENABLED");
            call.resolve(ret);
            return;
        }
        String query = call.getString("query", "").trim();
        if (query.isEmpty()) { call.reject("query is required"); return; }

        JSArray results = new JSArray();
        String[] projection = {
            ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME,
            ContactsContract.CommonDataKinds.Phone.NUMBER
        };
        String selection = ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME + " LIKE ?";
        String[] selectionArgs = { "%" + query + "%" };

        try (Cursor cursor = getContext().getContentResolver().query(
                ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
                projection, selection, selectionArgs,
                ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME + " ASC")) {
            if (cursor != null) {
                // The same phone number can appear more than once (synced from
                // multiple accounts) — dedupe so one contact isn't listed 3x.
                Set<String> seen = new HashSet<>();
                int nameIdx = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME);
                int numberIdx = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER);
                while (cursor.moveToNext() && results.length() < 5) {
                    String name = cursor.getString(nameIdx);
                    String number = cursor.getString(numberIdx);
                    if (name == null || number == null) continue;
                    String dedupeKey = name + "|" + number.replaceAll("[^0-9+]", "");
                    if (!seen.add(dedupeKey)) continue;
                    JSObject o = new JSObject();
                    o.put("name", name);
                    o.put("phone", number);
                    results.put(o);
                }
            }
        } catch (Exception e) {
            JSObject ret = new JSObject();
            ret.put("ok", false);
            ret.put("error", e.getMessage());
            call.resolve(ret);
            return;
        }

        JSObject ret = new JSObject();
        ret.put("ok", true);
        ret.put("results", results);
        call.resolve(ret);
    }
}

File: ios/App/App/VoidContactsPlugin.swift
import Foundation
import Capacitor
import Contacts

/**
 * Read-only contact lookup by name — the iOS counterpart to
 * android/.../VoidContactsPlugin.java. Same plugin name ("VoidContacts") and
 * same method shapes (isEnabled/requestAccess/find -> {ok, results:[{name,phone}]})
 * so app.js's contactsPlugin() helper works identically on both platforms
 * with zero JS changes. Never writes to or modifies contacts.
 */
@objc(VoidContactsPlugin)
public class VoidContactsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "VoidContactsPlugin"
    public let jsName = "VoidContacts"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestAccess", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "find", returnType: CAPPluginReturnPromise)
    ]

    @objc func isEnabled(_ call: CAPPluginCall) {
        let status = CNContactStore.authorizationStatus(for: .contacts)
        call.resolve(["value": status == .authorized])
    }

    @objc func requestAccess(_ call: CAPPluginCall) {
        let status = CNContactStore.authorizationStatus(for: .contacts)
        if status == .authorized {
            call.resolve()
            return
        }
        CNContactStore().requestAccess(for: .contacts) { granted, _ in
            if granted {
                call.resolve()
            } else {
                call.reject("CONTACTS_PERMISSION_DENIED")
            }
        }
    }

    /// query: a name (or partial name) to search for. Returns up to 5
    /// name+phone matches, contact order as returned by the Contacts store.
    @objc func find(_ call: CAPPluginCall) {
        let status = CNContactStore.authorizationStatus(for: .contacts)
        guard status == .authorized else {
            call.resolve(["ok": false, "error": "NOT_ENABLED"])
            return
        }
        let query = (call.getString("query") ?? "").trimmingCharacters(in: .whitespaces)
        guard !query.isEmpty else {
            call.reject("query is required")
            return
        }

        let store = CNContactStore()
        let keys = [CNContactGivenNameKey, CNContactFamilyNameKey, CNContactPhoneNumbersKey] as [CNKeyDescriptor]
        let request = CNContactFetchRequest(keysToFetch: keys)
        let lowerQuery = query.lowercased()
        var results: [[String: String]] = []
        // The same phone number can appear more than once per contact — dedupe
        // so one person isn't listed multiple times.
        var seen = Set<String>()

        do {
            try store.enumerateContacts(with: request) { contact, stop in
                if results.count >= 5 { stop.pointee = true; return }
                let fullName = "\(contact.givenName) \(contact.familyName)"
                    .trimmingCharacters(in: .whitespaces)
                guard fullName.lowercased().contains(lowerQuery) else { return }
                for phone in contact.phoneNumbers {
                    let number = phone.value.stringValue
                    let dedupeKey = "\(fullName)|\(number.filter { "0123456789+".contains($0) })"
                    guard seen.insert(dedupeKey).inserted else { continue }
                    results.append(["name": fullName, "phone": number])
                    if results.count >= 5 { break }
                }
            }
        } catch {
            call.resolve(["ok": false, "error": error.localizedDescription])
            return
        }

        call.resolve(["ok": true, "results": results])
    }
}
